### DISCUSSION TITLE
How to obtain the barycentric coordinates of the intersection?

### DISCUSSION THREAD

**Discussion by woAIxuexiSR** [mitsuba-renderer/mitsuba3#1195] [Q&A] [answered]
Hi there,

When I use `si = scene.ray_intersect(ray)` to get the intersection, `si.uv` seems to be the texture coordinates rather than the barycentric coordinates. 

Is there a way to obtain the barycentric coordinates?

**Comment by njroussel** [comment, marked as answer]
Hi @woAIxuexiSR 

Indeed, the `SurfaceInteraction3f` struct only holds barycentric coordinates in its `uv` field if the mesh is untextured. If the mesh is textured, then `uv` is the texture coordinate.
You could theoretically recompute the barycentric coordinates from the `si.p` and `si.prim_index` fields, but that would require access to the mesh's datastructures. 

If you need this information, I think the easiest way is to add an extra field to the `SurfaceInteraction3f` struct, fill it during the ray-intersection routing and apply that change wherever required.
